models/experimental/llama2_70b/tests/unit_tests/test_flash_sdpa.py

```diff
@@ -0,0 +1,429 @@
+    gt = torch.nn.functional.scaled_dot_product_attention(Q, K, V, attn_mask, is_causal=False)
+
+    mine = sdpa(Q, K, V, attn_mask)
+
+    out_pass, out_pcc = comp_pcc(gt, mine, 0.99)
+    print(f"python vs pytorch: {out_pcc}")
+    assert out_pass
+
+    fa = fa2(Q, K, V, attn_mask)
+    out_pass, out_pcc = comp_pcc(gt, fa, 0.99)
+    print(f"fa2 vs pytorch: {out_pcc}")
+    assert out_pass
+
+
+def run_test_sdpa_cb_python():
+    b = 1
+    nh = 8
+    nkv = 1
+    s = 2048
+    d = 128
+
+    Q = torch.randn(b, nh, s, d)
+    K = torch.randn(b, nkv, s, d)
+    V = torch.randn(b, nkv, s, d)
+    attn_mask = torch.full((s, s), torch.finfo(torch.float32).min)
+    attn_mask = torch.triu(attn_mask, diagonal=1).expand(b, nh, -1, -1)
+
+    # Print shapes of all inputs along with input names
+    print(f"Q: {Q.shape}")
+    print(f"K: {K.shape}")
+    print(f"V: {V.shape}")
+    print(f"attn_mask: {attn_mask.shape}")
+
+    # is_causal must be false if we specify an attention_mask
+    gt = torch.nn.functional.scaled_dot_product_attention(Q, K, V, attn_mask, is_causal=False)
+
+    mine = fa2_cb(Q, K, V, attn_mask)
+
+    out_pass, out_pcc = comp_pcc(gt, mine, 0.99)
+    print(f"python vs pytorch: {out_pcc}")
+    assert out_pass
+
+    fa = fa2(Q, K, V, attn_mask)
+    out_pass, out_pcc = comp_pcc(gt, fa, 0.99)
+    print(f"fa2 vs pytorch: {out_pcc}")
+    assert out_pass
+
+
+def run_stress_sdpa_tt(device):
+    b = 1
+    nh = 8
+    nkv = 1
+    s = 2048
+    d = 128
+
+    device.enable_program_cache()
+
+    program_config = tt_lib.operations.primary.transformers.SDPAMultiCoreProgramConfig(
+        compute_with_storage_grid_size=[8, 8], q_chunk_size=256, k_chunk_size=256
+    )
+
+    Q = torch.randn(b, nh, s, d)
+    K = torch.randn(b, nkv, s, d)
+    V = torch.randn(b, nkv, s, d)
+    attn_mask = torch.full((s, s), torch.finfo(torch.float32).min)
+    # attn_mask = torch.full((s, s), -10.0) #TODO: While exp is being debugged
+    attn_mask = torch.triu(attn_mask, diagonal=1).expand(b, 1, -1, -1)
+    # FOR DEBUG, don't use neginf
+
+    # Print shapes of all inputs along with input names
+    print(f"Q: {Q.shape}")
+    print(f"K: {K.shape}")
+    print(f"V: {V.shape}")
+    print(f"attn_mask: {attn_mask.shape}")
+
+    gt = torch.nn.functional.scaled_dot_product_attention(Q, K, V, attn_mask)
+
+    tt_q = torch2tt_tensor(Q, device)
+    tt_k = torch2tt_tensor(K, device)
+    tt_v = torch2tt_tensor(V, device)
+    tt_attn_mask = torch2tt_tensor(attn_mask, device)
+
+    first_pcc = 0.0
+    for i in range(1000):
+        print(f"Iteration {i}")
+        mine = tt_fa2_noconvert(device, tt_q, tt_k, tt_v, tt_attn_mask, program_config)
+        out_pass, out_pcc = comp_pcc(gt, mine, 0.99)
+        if i == 0:
+            first_pcc = out_pcc
+        else:
+            # assert out_pcc == first_pcc, "ND PCC found"
+            pass
+
+        print(f"python vs pytorch: {out_pcc}")
+        # assert out_pass
```

Comment:
Get rid of prints and replace with logger.debug for cleanliness.
For testing program cache, you should:
- Assert on number of expected caches (`device.num_program_cache_entries()`)
- Try to shift the memory of your inputs when you loop so you actually test the functionality of the callback